### PR TITLE
dev-env: Replace custom "dade" scripts with `use nix`.

### PR DESCRIPTION
**Pull request by SamirTalwar-DA**
You're going to hate me for this one. But look at the savings! :upside_down_face: 

This replaces our custom "dade" scripts with [`use nix`](https://github.com/direnv/direnv/wiki/Nix), the officially-sanctioned way to use Nix with direnv.

You can install [nix-direnv](https://github.com/nix-community/nix-direnv) to make it faster.

This PR doesn't go far enough, IMO. I think we can also clean up _nix/default.nix_ a lot. I tried to minimize it to only the changes necessary to get this working.

It's worth noting that this changes the behavior from lazy (install tools on demand) to eager (install all tools). It's pretty fast though, and I think worth it for the drop in complexity.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by robin-da** [comment]
:+1: . I never understood why we didn't do this from the start.

**Comment by SamirTalwar-DA** [comment]
I haven't forgotten about this. Hoping to pick it back up and figure out why the node.js Bazel rules are failing at some point soon.

**Comment by SamirTalwar-DA** [comment]
I'm getting this fun error again:

```
Not a regular file: /home/vsts/.cache/bazel/_bazel_vsts/d5c8b3f1185416e5207d53317d3a7876/external/node_nix/node_nix/bin/node
```

Weirdly, this didn't fail for [`2433279` (#7804)](https://github.com/digital-asset/daml/pull/7804/commits/24332793a4ecd907c97d8cc4b666e72056942e6e).

The rest works though. I have a feeling that if this runs on a fresh machine (with no disk cache), it should work, and then continue to work.

**Comment by garyverhaegen-da** [comment]
> You're going to hate me for this one. But look at the savings! 🙃

No hatred from me. I'd be thrilled if you could get it working _and_ get me to understand how it works and what the tradeoffs are.

> This PR doesn't go far enough, IMO. I think we can also clean up _nix/default.nix_ a lot. I tried to minimize it to only the changes necessary to get this working.

Baby steps are easier to review & merge. No need to go on a crusade and burn everything at once. For example:

- I think keeping `dev-env/bin` in `PATH` would make the transition easier. It looks like the switch allows us to remove most of the symlinks, but there are other things in `dev-env/bin`.
- I don't understand why you're trying to replace `dev-env/bin/dade-assist` with `dev-env/load.sh`. That seems unnecessarily invasive to me, especially considering that there may be other projects depending on this. (I know there were at some point builds of other projects that downloaded and built daml; not sure whether that's still the case, but still, some sort of transition period would be nice.) Why not just change what `dev-env/bin/dade-assist` does? Looks like something like:

```bash
set -euo pipefail
DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" &> /dev/null && pwd )"
echo "Warning: dade-assist is going away, please switch to dev-env/load.sh" >&2
cat $DIR/../load.sh
```

looks like it may preserve compatibility with minimal change to the current state of your PR.

> It's worth noting that this changes the behavior from lazy (install tools on demand) to eager (install all tools). It's pretty fast though, and I think worth it for the drop in complexity.

That's fine now, as we've removed most of our custom nix expressions.

**Comment by SamirTalwar-DA** [comment]
Thank you for the feedback, @garyverhaegen-da. I agree; I want to see this _work_ but I will not be opening this PR for review. Once it's functional, I will break it into several PRs which should each do less damage.

Good point about `dade assist`; let's make that easier to migrate.

I didn't really expect to get very far with this today; I just wanted to rebase it to see if it broke and I was surprised to see that it wasn't that difficult. I don't expect to get any of these changes in very soon. Just wondering if the branch was still useful.

File: ci/build.yml [hunk 1]
```diff
@@ -61,11 +61,27 @@ jobs:
   parameters:
     final_job_name: da_ghc_lib
 
+- job: build_dev_env
+  steps:
```
Comment (garyverhaegen-da) [inline ci/build.yml:65]:
```suggestion
  pool:
    name: ubuntu_20_04
    demands: assignment -equals default
  steps:
```

**Comment by realvictorprm** [comment]
This already convinces me that this change is good including the conversation I had with @garyverhaegen-da about this :smile: 

image

**Comment by SamirTalwar-DA** [comment]
It would be a lot better if it worked. :upside_down_face:

**Pull request closed without merging**